[ASan][libc++] std::basic_string annotations

AdvenamTacet opened the pull request:
This commit introduces basic annotations for `std::basic_string`, mirroring the approach used in `std::vector` and `std::deque`. Initially, only long strings with the default allocator will be annotated. Short strings (_SSO - short string optimization_) and strings with non-default allocators will be annotated in the near future, with separate commits dedicated to enabling them. The process will be similar to the workflow employed for enabling annotations in `std::deque`.

**Please note**: these annotations function effectively only when libc++ and libc++abi dylibs are instrumented (with ASan). This aligns with the prevailing behavior of Memory Sanitizer.

To avoid breaking everything, this commit also appends `_LIBCPP_INSTRUMENTED_WITH_ASAN` to `__config_site` whenever libc++ is compiled with ASan. If this macro is not defined, string annotations are not enabled. However, linking a binary that does **not** annotate strings with a dynamic library that annotates strings, is not permitted.

Originally proposed here: https://reviews.llvm.org/D132769

Related patches on Phabricator:
- Turning on annotations for short strings: https://reviews.llvm.org/D147680
- Turning on annotations for all allocators: https://reviews.llvm.org/D146214

This PR is a part of a series of patches extending AddressSanitizer C++ container overflow detection capabilities by adding annotations, similar to those existing in `std::vector` and `std::deque` collections. These enhancements empower ASan to effectively detect instances where the instrumented program attempts to access memory within a collection's internal allocation that remains unused. This includes cases where access occurs before or after the stored elements in `std::deque`, or between the `std::basic_string`'s size (including the null terminator) and capacity bounds.

The introduction of these annotations was spurred by a real-world software bug discovered by Trail of Bits, involving an out-of-bounds memory access during the comparison of two strings using the `std::equals` function. This function was taking iterators (`iter1_begin`, `iter1_end`, `iter2_begin`) to perform the comparison, using a custom comparison function. When the `iter1` object exceeded the length of `iter2`, an out-of-bounds read could occur on the `iter2` object. Container sanitization, upon enabling these annotations, would effectively identify and flag this potential vulnerability.

This Pull Request introduces basic annotations for `std::basic_string`. Long strings exhibit structural similarities to `std::vector` and will be annotated accordingly. Short strings are already implemented, but will be turned on separately in a forthcoming commit. Look at [a comment](https://github.com/llvm/llvm-project/pull/72677#issuecomment-1850554465) below to read about SSO issues at current moment.

Due to the functionality introduced in [D132522](https://github.com/llvm/llvm-project/commit/dd1b7b797a116eed588fd752fbe61d34deeb24e4), the `__sanitizer_annotate_contiguous_container` function now offers compatibility with all allocators. However, enabling this support will be done in a subsequent commit. For the time being, only strings with the default allocator will be annotated.

If you have any questions, please email:
- advenam.tacet@trailofbits.com
- disconnect3d@trailofbits.com

Commit:
[ASan][libc++] std::basic_string annotations

This commit turns on basic annotations for std::basic_string, similar to those in std::vector and std::deque.

Originally proposed here: https://reviews.llvm.org/D132769

Related patches:
Turning on annotations for short strings: https://reviews.llvm.org/D147680

Turning on annotations for all allocators: https://reviews.llvm.org/D146214

This revision is a part of a series of patches extending AddressSanitizer C++ container overflow detection capabilities by adding annotations, similar to those existing in std::vector, to std::string and std::deque collections. These changes allow ASan to detect cases when the instrumented program accesses memory which is internally allocated by the collection but is still not in-use (accesses before or after the stored elements for std::deque, or between the size and capacity bounds for std::string).

The motivation for the research and those changes was a bug, found by Trail of Bits, in a real code where an out-of-bounds read could happen as two strings were compared via a std::equals function that took iter1_begin, iter1_end, iter2_begin iterators (with a custom comparison function). When object iter1 was longer than iter2, read out-of-bounds on iter2 could happen. Container sanitization would detect it.

This Pull Request adds annotations for `std::basic_string`. Long string is very similar to std::vector, and therefore works well with `__sanitizer_annotate_contiguous_container` from LLVM 15 and therefore annotations works if that implementation is compiled with previous LLVM.
However, only the standard allocator is supported.

As D132522 extended possibilities of `__sanitizer_annotate_contiguous_container`, now annotations may be supported with all allocators, but that support will be added in a next PR. Only strings with default allocator are annotated with that commit.

If you have any questions, please email:
- advenam.tacet@trailofbits.com
- disconnect3d@trailofbits.com

AdvenamTacet commented:
@ldionne Given that the requirement for sanitized libc++ and libc++abi is already an established practice (Memory Sanitizer), could we proceed with merging this branch after addressing all code review concerns, prior to developing a unified approach for shipping instrumented libraries?

Maintaining this patch outside of upstream takes a lot of time due to fact how many files/functions are modified across the codebase. Each refactor/update/change of names on upstream requires dealing with merge conflicts because of that. Occasionally, identifying issues proves challenging without a thorough examination of all changes. Sometimes it's just hard to notice, unlucky conflict resolving problem. For instance, I recently spent hours fighting an issue caused by "auto resolving conflicts" duplicating a line of code, which resulted in difficult-to-detect undefined behavior. I would love to spend more time on developing something valuable for LLVM instead of maintaining the patch outside of upstream.

ldionne commented:
^ Yes, I think that's a reasonable approach. So we'll assume that libc++ is built with support for asan + container annotations if a user enables these container annotations in their code? Does this mean that `-fsanitize=address` will stop working out of the box with a non-asan `libc++.dylib`? That would be a problem.

AdvenamTacet commented:
Thank you!

> Does this mean that -fsanitize=address will stop working out of the box with a non-asan libc++.dylib? 

Now yes, but adding a macro to opt-in or opt-out is trivial, we already have the `#if` in place, it's enough to add a condition (`&& defined(ANNOTATE_STRING_OR_STH)`).

I'm in favor of opt-out macro, instead of opt-in. I'm aware of problems resulting from this decision, but using sanitized dylib should be the standard anyway and we should push in that direction.

We can always change opt-out to opt-in before branching, if we get feedback suggesting it.
Making it initially opt-out gives us more feedback now and the experience with previous containers shows that it's very important.

Let me know what version you prefer and I will add it.

PS
Ofc, not-annotated binary won't work with annotated libc++.dylib as well.

AdvenamTacet commented:
@philnik777 since your last review at Phabricator, except removing #ifdef around exported functions as requested there, I was doing only maintenance (`pull --rebase` etc.).

If you see nothing else to fix or only minor things, could you greenlight the review?

AdvenamTacet commented:
@ldionne I'd appreciate a review of that PR, please let me know if you have time to look at it anytime soon.

EricWF reviewed (changes requested):
Could you please write specific tests for your change. 

I don't think tacking on an assertion to every existing test is the right way to go. I understand the appeal, but I think specific tests written expressly to test this change is the right way to go.

In the past, we've allowed tests to be added like this, but it's resulted in tests that run too long and tests which don't test a single thing.

AdvenamTacet commented:
Hey @EricWF , that way of testing ASan is already established in `std::deque` and `std::vector`.
Overhead for tests runtime exists only when those tests are compiled with ASan (therefore, only the `generic-asan` build) plus the overhead for `safe_allocator` tests (basically `min_allocator` with cleaning memory).

We did the same thing with `std::deque` and certain bugs remained undetected before upstreaming, what suggests that ASan tests cannot be less extensive.

Creating a separate set of tests is not  feasible, as it would basically require copy-pasting all tests we have, as almost every function has to be carefully tested. And if a new function is added/overloaded in future, tests should be added in two places.

To address performance concerns when running tests with `safe_allocator`, I can replace `min_allocator` tests with `safe_allocator`, as the latter tests everything what the former and extends beyond it (instead of duplicating tests by adding same ones with the `safe_allocator`, I can replace all instances of `min_allocator` with `safe_allocator` wherever `safe_allocator` is used).
If we don't want two tests, I'm in favor of replacing `min_allocator` with `safe_allocators`, but another approach is to guard `safe_allocator` tests by an ifdef to ensure execution only when compiled with AddressSanitizer. 

What do you think?

Commit:
Refactor: remove duplicated code with `[[__maybe_unused__]]`

This commit refactors the ASan annotation function to reduce unnecessary code duplication.

Suggested here: https://github.com/llvm/llvm-project/issues/73043
Related PR: https://github.com/llvm/llvm-project/pull/74023

ldionne reviewed (changes requested):
As-is, landing this patch will break most users who use `-fsanitize=address` and use libc++ in their code. Indeed, vendors ship a single version of `libc++.dylib` right now and that version is not instrumented, which means that the extern template instantiations in `libc++.dylib` will cause problems.

I'm worried that this will be a significant regression. Ideally, we'd address the issue by having a proper plan for shipping an instrumented library and having the driver select it based on `-fsanitize=address`. But barring that, we'll need to either
1. make the `string` annotations opt-in, or
2. disable the extern template instantiation declarations under asan
3. only enable these annotations when the dylib was built with support for ASAN

(1) basically negates the benefit of this patch since almost nobody will turn it on
(2) is still not correct because of ODR violations, but it should make most basic use cases work. It would still be a temporary solution.

I think (3) is probably what makes the most sense, since that wouldn't be a temporary solution. It would only mean that most platforms don't get this support by default because the dylib isn't built with instrumentation, but the minute we ship a library that does support that, everything would still work out of the box.

This could be done by encoding whether ASAN container annotations are supported in the `__config_site`, and that would then become a configuration-time property of the library.

ldionne commented:
If we encode the fact that libc++ was built with instrumentation in `__config_site`, a user should be able to pass `-fsanitize=address` while building against a library that isn't built with annotations, and the container annotations would be disabled. And if a user uses a dylib that is built with instrumentation and doesn't pass `-fsanitize=address`, I'm not sure what would happen but we should make it work.

Basically the idea would be to make these container annotations distinct from whether the user is compiling with `-fsanitize=address`, and in the future we should ensure that we ship an instrumented version of libc++ at all times, and then make `-fsanitize=address` select that instrumented version.

AdvenamTacet commented:
> And if a user uses a dylib that is built with instrumentation and doesn't pass -fsanitize=address, I'm not sure what would happen but we should make it work.

I don't think we can make it work, linking is going to fail in that case. It already cannot work, string annotations have no impact on it.

> 2. disable the extern template instantiation declarations under asan
> [...]
> (2) is still not correct because of ODR violations, but it should make most basic use cases work. It would still be a temporary solution.

While I agree, it's far from perfect, we already assume that ASan has no stable ABI and when compiling with ASan we should use the same version of libc++ and LLVM.
Therefore, temporarily (2) plus opt-out makro for less basic cases with non-instrumented dylib seem good for me.
A significant advantage of this solution is that all users will now have annotated strings.

But (3) sounds ok as well, if we just have a makro in `__config_site`, it's future-proof and easy to implement in libc++. 
Biggest downside is fact that users don't compile libc++ with ASan already, they have to be aware of that change to benefit from it, until we figure out how we can automatically choose a dylib based on `-fsanitize=address` flag.

@ldionne should we go with(3)?

AdvenamTacet commented on libcxx/test/std/strings/basic.string/string.cons/initializer_list.pass.cpp:
```diff
@@ -18,18 +18,27 @@
+    LIBCPP_ASSERT(is_string_asan_correct(s));
+  }
+  {
+    std::basic_string<char, std::char_traits<char>, Alloc<char> > s;
+    s = {'a', 'b', 'c'};
+    assert(s == "abc");
+    LIBCPP_ASSERT(is_string_asan_correct(s));
   }
```
A similar test already exists in the correct place, so I simply removed that one.

AdvenamTacet commented on libcxx/test/std/strings/basic.string/string.modifiers/string_erase/pop_back.asan.pass.cpp:
```diff
@@ -0,0 +1,50 @@
+//
+//===----------------------------------------------------------------------===//
+
+// UNSUPPORTED: c++03
```
This test is covered in the main file, so I removed it.

Commit:
Address code review feedback for existing code

Requested changes for existing code.
Additional changes related to dylibs in the next commit.

This commit extends a few more tests to cover long strings in more situations.

Commit:
Add flag informing about dylib ASan instrumentation and guard string annotations

* CMake files are updated to automatically add the `_LIBCPP_INSTRUMENTED_WITH_ASAN`
macro to `__config_site`. This macro is used to inform the compiler (in header files)
that the library has been built with ASan and is instrumented for memory safety checks.

* The `-D_LIBCPP_INSTRUMENTED_WITH_ASAN` flag is added to the libc++ compilation
command when building with ASan. This flag ensures that basic_string annotations
are included in the binary, which are needed for ASan to track memory usage
correctly.

* The string file is updated to use string annotations from headers if and only
if the code is compiled with `-fsanitize=address` and the dylib is compiled
with ASan (so, `_LIBCPP_INSTRUMENTED_WITH_ASAN` is defined). This ensures that
the string annotations are only included when they are supported.

Commit:
Add missing annotations checks

Some files are missing annotations checks, this commit adds those checks.

Commit:
Additional test cases

This commit adds a few test cases to check ASan annotations for SSO/no-SSO.

Commit:
[ASan] Annotations fix

Resolved an issue with ASan annotations that was causing failure of `substr_rvalue` tests.

Commit:
[ASan] SSO annotations fixes and refactor

This commit fixes issues related to annotations and instrumentation, causing problems only with SSO annotations.

First, it fixes an issue where `_LIBCPP_STRING_INTERNAL_MEMORY_ACCESS` doesn't work in `basic_string<_CharT, _Traits, _Allocator>::__move_assign(basic_string& __str, true_type)``.

Second, it adds missing SSO annotations and refactors annotations around `__uninitialized_size_tag`.

To reproduce the errors in this commit, change `_LIBCPP_SHORT_STRING_ANNOTATIONS_ALLOWED` to true. This flag will be changed to true (and probably deleted) soon.

Commit:
[ASan] Cleaning _LIBCPP_STRING_INTERNAL_MEMORY_ACCESS

EricWF reviewed (approved):
LGTM with the understanding that there's some iteration left to be done.
I think it's reasonable to want to land this before the world is formatted.

AdvenamTacet commented:
Ok, I made (hopefully) the final iteration, where I  [removed all unnecessary `_LIBCPP_STRING_INTERNAL_MEMORY_ACCESS`](https://github.com/llvm/llvm-project/pull/72677/commits/96f1712418457f21268db22f3b6feaed4b810abc).

This PR does [not turn on short string annotations](https://github.com/llvm/llvm-project/pull/72677/files#diff-534bc2907ddb3b074ded1353d18fd7d578daf1707943b3039bab4ed975aba3b3R662).
I'm not aware of any issues with that implementation.

I tested everything locally with SSO annotations turned on, to have working implementation, in which we can just switch off to on, without any other code modification.

However, right now turning on SSO annotations is not possible due to a [very specific implementation of sstream](https://github.com/llvm/llvm-project/blob/main/libcxx/include/sstream#L412), where `__move_assign` is used. That affects only short string annotations and should not block this PR. I'm going to address this in a separate PR.

@EricWF thank you for the code review and feedback on Discord!

@ldionne if possible, could we please merge it before formatting lands? If you want me to change something, I'm happy to sit to it asap.

ldionne commented on libcxx/CMakeLists.txt:
```diff
@@ -543,6 +543,18 @@ function(cxx_add_basic_build_flags target)
   # library.
   target_compile_definitions(${target} PRIVATE -D_LIBCPP_BUILDING_LIBRARY)
 
+  # Define _LIBCPP_INSTRUMENTED_WITH_ASAN while building the library with ASan.
+  # Normally, the _LIBCPP_INSTRUMENTED_WITH_ASAN flag is used to keep information whether
+  # dylibs are built with AddressSanitizer. However, when building libc++,
+  # this flag needs to be defined so that the resulting dylib has all ASan functionalities normally guarded by this flag.
+  # If the _LIBCPP_INSTRUMENTED_WITH_ASAN flag is not defined, then parts of the ASan instrumentation code in libc++
+  # will not be compiled into it resulting in true positives.
+  # For context, read: https://github.com/llvm/llvm-project/pull/72677#pullrequestreview-1765402800
+  string(FIND "${LLVM_USE_SANITIZER}" "Address" building_with_asan)
+  if (NOT "${building_with_asan}" STREQUAL "-1")
+    target_compile_definitions(${target} PRIVATE -D_LIBCPP_INSTRUMENTED_WITH_ASAN)
+  endif()
```
You don't need this part since you are encoding it in the `__config_site` header. You can remove this but move the comment to below.

AdvenamTacet replied:
But from what I understand, `__config_site` is not available while building libc++ dylibe, that it.

AdvenamTacet replied:
I see how it should work, confirming that it really works.

ldionne commented on libcxx/CMakeLists.txt:
```diff
@@ -642,6 +654,13 @@ get_sanitizer_flags(SANITIZER_FLAGS "${LLVM_USE_SANITIZER}")
+# _LIBCPP_INSTRUMENTED_WITH_ASAN informs that library was built with ASan.
+string(FIND "${LLVM_USE_SANITIZER}" "Address" building_with_asan)
+if (NOT "${building_with_asan}" STREQUAL "-1")
+  set(LIBCPP_SANITIZER_DEFINES "#define _LIBCPP_INSTRUMENTED_WITH_ASAN 1")
+  config_define(${LIBCPP_SANITIZER_DEFINES} _LIBCPP_SANITIZER_DEFINES)
```
Let's just do this instead:

```
config_define(ON _LIBCPP_INSTRUMENTED_WITH_ASAN)
```

Then you don't need `@_LIBCPP_SANITIZER_DEFINES@` below at all.

ldionne commented on libcxx/include/__config_site.in:
```diff
@@ -45,6 +45,7 @@
 #endif
 
 @_LIBCPP_ABI_DEFINES@
+@_LIBCPP_SANITIZER_DEFINES@
```
You can remove this and instead use `#cmakedefine _LIBCPP_INSTRUMENTED_WITH_ASAN` above, right below `_LIBCPP_HAS_NO_TIME_ZONE_DATABASE`.

ldionne commented on libcxxabi/CMakeLists.txt:
```diff
@@ -269,6 +269,13 @@ add_definitions(-D_LIBCXXABI_BUILDING_LIBRARY)
 # it is being built as part of libcxx.
 add_definitions(-D_LIBCPP_BUILDING_LIBRARY)
 
+# To ensure that all ASan functionalities are built into the binary.
```
This also shouldn't be necessary anymore with my suggested `__config_site` changes.

ldionne reviewed (commented):
This LGTM modulo a few comments. I'll double-check the PR again after you've applied them but this should be ready. Thanks for the patch and for your display of patience with us during this review -- I thought this would be really problematic until I realized we could use `__config_site` to solve most of the problems I saw with the patch initially.

Commit:
Updating CMake files related to `__config_site`

Based on code review feedback.

AdvenamTacet commented:
@ldionne ping. I implemented requested changes, locally everything seems to work. I'm happy to merge this  PR as soon as you accept the review.

ldionne reviewed (approved):
LGTM w/ green CI.

vitalybuka commented:
This breaks this bot https://lab.llvm.org/buildbot/#/builders/5/builds/39229
